WIP: Implement a SPIR-V linker

pierremoreau opened the pull request:
Still a work-in-progress and I can most likely reuse more code, but at least I adapted the code to use some of the SPIRV-Tools interface/code.


Missing:

- [x] Update SPIR-V version support from 1.0 to 1.2
- [X] Remove duplicate types
- [x] Remove duplicate decorations
- [ ] Remove duplicate names
- [X] Remove duplicate categories
- [x] Remove duplicate `SpvOpExtInstImport`
- [x] Tests
- [x] Ensure that an import/export pair of variables or functions have the same interface before accepting to link

Questions:

- Should the linker require the given binaries to be valid SPIR-V binaries?
- Should the linker require the given binaries to have the same endianness as the host?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/SPIRV-Tools?pullRequest=693) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/SPIRV-Tools?pullRequest=693) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/SPIRV-Tools?pullRequest=693) it.</sub>

pierremoreau commented:
I am a bit confused with those errors from travis: how can variables local to a lambda be shadowing variables declared outside of, and not captured by, the lambda?

dneto0 commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+  spv_result_t ret;
+  uint32_t max_version = 0u;
+
+  // As we need to bump IDs in each module, so as they do not conflict as we
```
Please give a high level description.  
Phase 1: Shift the Ids in each binary so they occupy disjoint ranges.  In fact, I recommend you refactor out this transform step into its own function, so the high level structure is easier to see.
Phase 2: Merge the binaries.

pierremoreau replied:
Sounds like a good idea.

dneto0 commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+    if (ret != SPV_SUCCESS)
+      return ret;
+
+    id_upper_bound += binary_max_id;
```
Please error out if max Id goes beyond the unviersal limit of 0x3FFFFF

pierremoreau replied:
I had forgotten IDs had a limit of 0x3FFFFF, I will add that.

dneto0 commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+
+  // id_upper_bound is currently equal to the highest id being used, so add
+  // one to get the strict upper bound
+  ++id_upper_bound;
```
Error out if this exceeds 0x3FFFFF

pierremoreau replied:
Will od

dneto0 commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+  case spv::AddressingModel::Physical64: return "Physical64";
+  case spv::AddressingModel::Max:        return "Unsupported";
+  }
+}
```
Actually, this should be made grammar-driven.  
Use the context to create an AssemblyGrammar object, then do 
spv_operand_desc desc = nullptr;
lookupOperand(SPV_OPERAND_TYPE_ADDRESSING_MODEL, uint32_t(model), &desc);
then use desc->name

pierremoreau replied:
I will have a look at how to generate those grammars, and replace this method and the one below with that.

dneto0 commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+  case spv::MemoryModel::GLSL450: return "GLSL450";
+  case spv::MemoryModel::OpenCL:  return "OpenCL";
+  case spv::MemoryModel::Max:     return "Unsupported";
+  }
```
Similar to above, do a lookupOperand to get the name, but this time with  SPV_OPERAND_TYPE_MEMORY_MODEL

dneto0 commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+
+using spirv_word = uint32_t;
+
+spirv_word GetWord(const uint32_t* binary, size_t word_offset) {
```
Style guide requires method comments on every function and method.

pierremoreau replied:
I will keep that in mind and add comments to all new functions I create, as well as to older ones that are missing one.

dneto0 commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+  return reinterpret_cast<const char*>(binary + word_offset);
+}
+
+/// Extracts the opcode and the number of words making up this instruction.
```
If you use the binary parser then this is decomposed for you. (see below for higher level feedback at TransformBinary)

pierremoreau replied:
Ack’ed

dneto0 commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+  }
+}
+
+static spv_result_t MergeModules(spv_context_t& context,
```
There's a lot of custom code here to sort out instructions into their own sections.  I think you could shorten and simplify a lot of this code by reusing the Module class from the optimizer.  See source/opt/module.h.

Its internal representation already segregate the instructions in different logical layout sections.  You make a Module with BuildModule from opt/build_module.h.
There are already iterator ranges for things like debug instructions, types, annotations, entry_points.  I would encourage adding methods to get iterator ranges for the other layout sections.

pierremoreau replied:
I will definitely look into that class, thanks for the pointer!

I was thinking, while submitting an earlier version to Mesa, in making a small metadata structure that would hold where each section started, to avoid all this checking and splitting, as it is annoying and the current code does not really highlight on which section it is currently working on.

dneto0 replied:
At the moment, the optimizer classes are in a separate library.
 
However, experience with the validator (in the main library) and the optimizer shows they often need the same things, but have different internal representations of the SPIR-V instructions.   I think the long term direction is to merge those representations and restructure the libraries.

For now, I suggest putting the linker entry points into the optimizer.hpp header.  At least it would correctly signal the library dependence.

pierremoreau commented on include/spirv-tools/libspirv.h:
```diff
@@ -271,6 +271,12 @@ typedef enum spv_binary_to_text_options_t {
 
+typedef enum spv_linker_options_t {
+  SPV_LINKER_OPTION_NONE = SPV_BIT(0),
+  SPV_LINKER_OPTION_CREATE_LIBRARY = SPV_BIT(1),
```
I’ll switch it to an opaque object, you are right.
If creating a library, I keep exported symbols, otherwise I remove them as we are producing an executable; I I have been mainly looking at it from an OpenCL’s viewpoint. I wanted to add more options, mainly the ones that can be passed to OpenCL in the linking step, like relaxing some conditions on floating point operations, but I haven’t looked into it yet.

pierremoreau commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+// Copyright (c) 2017 Pierre Moreau
```
I realised that by looking at the output from Travis. I’ll push a fix.

pierremoreau commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+  CopySections(IsExtInstImportOpcode);
+
+  // Merge OpMemoryModel
+  spv::AddressingModel addressing_model, addressing_model_tmp;
```
I will do that. I should probably use `spv::AddressModel::Max` and the memory model equivalent for the initialisation, as I don’t see what could be a good default value. And if after processing the binaries we still have the default values, then something wrong happened.

pierremoreau commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+    const size_t length = binaries[i].size();
+    const uint32_t index = indices[i];
+    uint32_t j = index;
+    uint32_t prev_j = j;
```
That is very true… a leftover from copy/pasting that loop structure from somewhere else. I am a bit surprised that clang did not pick it up locally. Similarly to all the errors/warning thrown by Travis. I’ll add locally some more warning flags to the default ones./

pierremoreau commented on source/linker.cpp:
```diff
@@ -0,0 +1,1098 @@
+}
+
+/// Applies the given function to all IDs found in the given binary.
+static spv_result_t TransformBinary(spv_context_t& context,
```
Tracking the width of the selector for the switch statements was a bit annoying, but not too difficult either.

Both approaches are interesting, and the binary parser having been fuzzed is a good plus.

pierremoreau commented on include/spirv-tools/libspirv.hpp:
```diff
@@ -117,6 +119,13 @@ class SpirvTools {
 
+  bool Link(const std::vector<std::vector<uint32_t>>& binaries,
+            std::vector<uint32_t>& linked_binary,
+            uint32_t options = kDefaultLinkerOptions) const;
```
Why should I drop the options argument from the C++ version of the functions? Is there another way to pass the linking options? (I mostly followed the API used by the assembler and disassembler.)

pierremoreau commented on include/spirv-tools/libspirv.hpp:
```diff
@@ -117,6 +119,13 @@ class SpirvTools {
   bool Validate(const uint32_t* binary, size_t binary_size,
                 const ValidatorOptions& options) const;
 
+  bool Link(const std::vector<std::vector<uint32_t>>& binaries,
```
Will add

pierremoreau commented on include/spirv-tools/libspirv.h:
```diff
@@ -474,6 +480,19 @@ spv_result_t spvValidateBinary(const spv_const_context context,
                                const uint32_t* words, const size_t num_words,
                                spv_diagnostic* diagnostic);
 
+spv_result_t spvLink(const spv_const_context context,
```
* Regarding the versions, I was assuming that a SPIR-V 1.0 binary could be marked as 1.1, except if instructions are deprecated, or their format changed but the latter should never happen (I guess). Therefore, I was making the linked binary use the maximum of all encountered versions among the binaries given as input.
* I was assuming that the input binaries had already been validated, to avoid having validation checks in the linker code. However, if I use `spvBinaryParse` or the IR from the optimiser, that might not be a problem.
  I guess validation would fail for a Vulkan module as Linkage requires Addresses, which itself requires a non-logical memory model, which is not supported by the Vulkan model?

pierremoreau commented:
Thank you very much for your review! There is quite a bit left to do, but it should result in code that is easier to read and maintain in the end, so it is totally worth it.

I have never been satisfied with the function transforming IDs, and I will be really happy to see it replaced.

pierremoreau commented:
I have been playing a bit with the IR from the optimiser, and it looks quite promising! I need to finish porting the code to it, and clean up everything.

dneto0 commented on source/link/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+
+set_property(TARGET SPIRV-Tools-link PROPERTY FOLDER "SPIRV-Tools libraries")
+
+install(TARGETS SPIRV-Tools-link
```
Please make installation conditional on if(ENABLE_SPIRV_TOOLS_INSTALL)
This is new since you started this PR.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
-};
-
-SpirvLinker::SpirvLinker(spv_target_env env) : impl_(new Impl(env)) {}
+  // TODO recompact IDs
```
Optional.  Also, there's an optimization pass for that.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+  ModuleHeader header;
+  header.magic_number = spv::MagicNumber;
+  header.version = spv::Version; // FIXME
+  header.generator = 0u;
```
Officially, 0 is reserved.  It's debatable whether we should reserve another for SPIRV-Tools optimizer & linker.
Please add a TODO to revisit this.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+
+  uint32_t id_bound = 0u;
+  for (const auto& mod : binaries) {
+    std::unique_ptr<Module> module = BuildModule(impl_->context->target_env, impl_->context->consumer, mod.data(), mod.size());
```
Suggest reformat with clang-format.  This will wrap for easier reading.

pierremoreau replied:
I will do that. I was planning to finish cleaning things up, then rebase on top of master, before pinging you for another round of reviews, but you were faster. :-D

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+  for (const auto& mod : binaries) {
+    std::unique_ptr<Module> module = BuildModule(impl_->context->target_env, impl_->context->consumer, mod.data(), mod.size());
+    if (module == nullptr)
+      return 1;
```
May as well return true.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+      return 1;
+    module->ForEachInst([&id_bound](Instruction* insn){
+        for (auto o = insn->begin(); o != insn->end(); ++o) {
+          if (!spvIsIdType(o->type))
```
Slightly prefer expressing this in the positive:

```
if (spvIsIdType(o->type)) o->words[0] += id_bound;
```

pierremoreau replied:
Will change it

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+    if (module == nullptr)
+      return 1;
+    module->ForEachInst([&id_bound](Instruction* insn){
+        for (auto o = insn->begin(); o != insn->end(); ++o) {
```
Range for reads better:
```
for (auto& operand : *insn) {
}
```

pierremoreau replied:
I think there was a reason why I did not went with the for-range approach, but I can’t remember it. I’ll have another look.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+  std::vector<std::unique_ptr<Module>> modules;
+  modules.reserve(binaries.size());
+
+  uint32_t id_bound = 0u;
```
Worth a short comment:  Shift IDs in each module so no ID ranges overlap.  Also compute the maximum ID.

pierremoreau replied:
I added some comments in the function to better mark the different steps, as well as turn some steps into a separate function.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
-SpirvLinker::~SpirvLinker() {}
+  ModuleHeader header;
+  header.magic_number = spv::MagicNumber;
+  header.version = spv::Version; // FIXME
```
Please use TODO(your-name) rather than FIXME.
Also, what needs fixing here?  Say you've got a decision to make.  Use the max version of any module you've seen?  I think that's probably right.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+    for (const auto& insn : module->ext_inst_imports())
+      linkedModule->AddExtInstImport(MakeUnique<Instruction>(insn));
+
+  // FIXME
```
Use TODO(your-name) rather than FIXME. 
Also say what the issue is.   I  guess it would be that it's an error to mix different memory models? Or is there a promotion lattice you can use?  Ok to error out for now(!)

pierremoreau replied:
That FIXME (and the other ones) are only for me to remember I need to come back to it before this gets merged.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+  for (const auto& module : modules)
+    linkedModule->SetMemoryModel(MakeUnique<Instruction>(*module->GetMemoryModel()));
+
+  for (const auto& module : modules)
```
Add TODO: error out if there are duplicate entry point names.

pierremoreau replied:
That is a good point! I should check on global variables and functions as well.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+    for (const auto& insn : module->annotations())
+      linkedModule->AddAnnotationInst(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules)
```
Add TODO: Must make unique non-aggregate non-pointer types.  (Pointer types now can be non-unique in SPV_KHR_variable_pointers)

pierremoreau replied:
Ah, I will have a look at the extension and add a todo.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+    for (const auto& insn : module->types_values())
+      linkedModule->AddType(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules) {
```
Worth a high level comment.   Process function declarations and bodies.

pierremoreau replied:
Wil add

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+      linkedModule->AddType(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules) {
+    for (auto i = module->cbegin(); i != module->cend(); ++i) {
```
Can we use range-for here?

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+      i->ForEachParam([&func](const Instruction* insn){
+          func->AddParameter(MakeUnique<Instruction>(*insn));
+      }, true);
+      for (auto j = i->cbegin(); j != i->cend(); ++j) {
```
Can we use range for here?

dneto0 commented on include/spirv-tools/linker.hpp:
```diff
@@ -0,0 +1,67 @@
+  // invoked once for each message communicated from the library.
+  void SetMessageConsumer(MessageConsumer consumer);
+
+  bool Link(const std::vector<std::vector<uint32_t>>& binaries,
```
Need method comment.  What is the return value?
For boolean return, we usually have true meaning success.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,292 @@
+  SetContextMessageConsumer(impl_->context, std::move(consumer));
+}
+
+bool SpirvLinker::Link(const std::vector<std::vector<uint32_t>>& binaries,
```
Please make return true be the success case.

pierremoreau replied:
I had changed the type to `spv_result_t` but those changes were not pushed yet.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,292 @@
+  for (const auto& mod : binaries) {
+    std::unique_ptr<Module> module = BuildModule(impl_->context->target_env, impl_->context->consumer, mod.data(), mod.size());
+    if (module == nullptr)
+      return 1;
```
Should be a failure case.
Should emit a diagnostic in failure cases.

We have a coding pattern of doing both diagnostic emission and returning of scalar result.  See use of diag() in source/val/validation_state.h, or uses of AssemblyContext::diagnostic in the assembler (source/text.cpp).

pierremoreau replied:
Same thing here, I made the change but did not push it.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
-    }
-
-    indices[i] = index + 3u;
+  std::unordered_map<spv::Id, std::string> imports;
```
Need a high level comment:  Resolve linkage imports and exports.

pierremoreau replied:
Fixed

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+    else if (linkage == spv::LinkageType::Export)
+      exports.emplace(reinterpret_cast<const char*>(insn.GetOperand(2u).words.data()), insn.GetSingleWordOperand(0u));
+    else
+      return 1;
```
Fail with diagnostic.

pierremoreau replied:
Fixed

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
+  for (const auto& i : imports) {
+    auto j = exports.find(i.second);
+    if (j == exports.end()) {
+      std::cerr << "Failed to find \"" << i.second << "\"\n";
```
Diagnostic through the context object.  You'll be able to write sensible tests that way.

pierremoreau replied:
Fixed

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
-      }
+  // Remove declarations of imported global variables
+  for (auto i = linkedModule->types_values_begin(); i != linkedModule->types_values_end();) {
+    if (i->NumOperands() < 2u) {
```
This is a bit mysterious.    Better to just compare opcode to OpVariable.

Suggest
```
if (i->opcode() == OpVariable) { .... do stuff ... }
```

I tend to prefer avoiding non-local control flow like continue and break when the straightforward expression is slightly more compact.  Framing in the positive is nicer, provided indentation remains sensibly bounded.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
-                (length - index) * sizeof(spirv_word));
-    merged_length += length - index;
-  }
+  // TODO Remove duplicate decorations
```
Explain more?  These would be decorations that are the same but on both an import ID and export ID that are really the same object?

dneto0 commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
-    return ret;
+  // Remove import linkage attributes
+  for (auto i = linkedModule->annotation_begin(); i != linkedModule->annotation_end();) {
+    if (i->opcode() != SpvOpDecorate ||
```
What if the import was not resolved?  Should only do this for imports in the linkage table.

pierremoreau replied:
If the import was not resolved, we wouldn’t be here as the code exits [here](https://github.com/KhronosGroup/SPIRV-Tools/pull/693/files/a45b072e2d85afb1193f8d9e8f879e4a5792b921..5be189f9291d60751246ede2f3b27400a14e151f#diff-202f9912f4c3d52d3531ffa6c2cbae74R170)

dneto0 commented:
Also needs a rebase so the bots can run.

pierremoreau commented on source/link/linker.cpp:
```diff
@@ -16,1183 +16,277 @@
-
-    position.index = i += insn_word_count;
-  }
+} // anonymous namespace
```
Same as above. I was very surprised initially to read your comment about deleting it, as I was 100% sure I had already done that.

pierremoreau commented:
I’ll address the other points over the week-end, among others formatting the code, missing comments and the options type.

Thanks again for reviewing!

dneto0 commented on source/link/linker.cpp:
```diff
@@ -157,6 +157,10 @@ spv_result_t Linker::Link(const std::vector<std::vector<uint32_t>>& binaries,
       for (auto& o : *insn)
         if (spvIsIdType(o.type)) o.words[0] += id_bound;
+      uint32_t result_id = insn->result_id();
+      if (result_id != 0u) insn->SetResultId(result_id + id_bound);
```
Optional:  A little coding trick.  You can push the definition of result_id into the conditional and make the check against 0 implicit.   

```
if (auto result_id = insn->result_id()) 
   insn->SetResultId(result_id + id_bound);
```

The chief advantages are: one less line of code to read, and the result_id variable goes out of scope right away, so you can't misuse it too much.

If the lambda's body were much longer, I'd ask for result_id and result_type to be made const, because they can.

pierremoreau replied:
Fixed and made const.

AlexeySotkin commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,348 @@
+  // Phase 1: Shift the IDs used in each binary so that they occupy a disjoint
+  //          range from the other binaries, and compute the new ID bound.
+  uint32_t id_bound = 0u;
+  for (auto& module : modules) {
```
I think the first iteration can be skipped.

pierremoreau replied:
Indeed it can, I’ll fix that.

pierremoreau commented:
This should fix the test regressions I introduced when splitting `debugs` into two sets.
I also removed the modifications I made to `libspirv.h` and `libspirv.hpp` as they are no longer needed, and rebased on master.

Next are:
* remove “duplicate” types
* add tests

I’ll probably leave linking flags other than “create a library” for a later pull request; I need to check some OpenCL behaviour first.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+
+  //Find the import/export pairs
+  std::vector<std::pair<LinkingData, LinkingData>> linkingsToDo;
+  for (const auto& i : modulesImports) {
```
Please rename i to something more descriptive.  Maybe "moduleImports".     (though that's a bit tricky with just making "modules" singular...)

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+  for (const auto& i : modulesImports) {
+    for (const auto& j : i) {
+      std::vector<LinkingData> possibleExports;
+      for (const auto& k : modulesExports) {
```
Suggest rename k --> moduleExports

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+    for (const auto& j : i) {
+      std::vector<LinkingData> possibleExports;
+      for (const auto& k : modulesExports) {
+        const auto& l = k.find(j.name);
```
Suggest rename  l --> where

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+    }
+  }
+
+  // Grab all types used by imports and exports
```
Reading this far, I don't know why this is necessary.  Add a comment saying why you need to do this.
Is it to catch a user error of trying to mismatched types?

Also, need a comment saying "now we fill in the typeId member of each linkage entry. (if I read correctly up until now)

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+  }
+
+  //Find the import/export pairs
+  std::vector<std::pair<LinkingData, LinkingData>> linkingsToDo;
```
Would really prefer a struct with sensible member names.  Hard to remember that first is the import and second is the export.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+  }
+
+  // Grab all types used by imports and exports
+  std::unordered_map<size_t, std::unordered_set<SpvId>> typesToSearchFor;
```
Describe this data structure.  What is it mapipng?

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+  // Grab all types used by imports and exports
+  std::unordered_map<size_t, std::unordered_set<SpvId>> typesToSearchFor;
+  for (auto& i : linkingsToDo) {
+    for (const auto& j : modules[i.first.moduleIndex]->types_values()) {
```
The "i" and "j" make it hard to read.  They are particular things:

 i --> importExportPair  
j --> importId

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+  for (auto& i : linkingsToDo) {
+    for (const auto& j : modules[i.first.moduleIndex]->types_values()) {
+      if (j.result_id() == i.first.id) {
+        assert(j.opcode() == SpvOpVariable);
```
So you only handle import/export of variables?
Need a TODO comment saying it should be extended to import/export of functions.
This also should be noted as a limitation at the API level and the command line help.

pierremoreau replied:
I am not, but it is not that easy to understand it from the code without any comments.

At that point in the code, I only know the IDs of imported/exported variables/functions, but I don’t know whether the ID corresponds to a variable or a function, so I start by looping over all variables. If at the end of the loop I still haven’t found the ID, I then loop over all the functions (lines 525–531).

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+    if (modules[i.first]->types_values().empty())
+      continue;
+
+    auto typesIter = modules[i.first]->types_values_end();
```
Say why you're going backward.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+    assert(i.second.typeId != 0u);
+  }
+  std::unordered_map<SpvId, TypeData> types;
+  for (auto& i : typesToSearchFor) {
```
What are you trying to do?   Need a comment.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+    } while (typesIter != modules[i.first]->types_values_begin());
+  }
+
+  // Ensure the import and export types are similar
```
Oh.  Should have said this long ago, before you started looking for type info.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+                 << " and exported function %" << i.second.id << "'s " << j << "th parameter.";
+  }
+
+  // TODO(pierremoreau): Remove import types
```
I think you just delegate this to the RemoveDuplicatesPass.  But you should say so here anyway.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+  }
+
+  // Remove prototypes of imported functions
+  for (const auto& i : linkingsToDo) {
```
These might have OpName attached to them.
Add TODO for their removal.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+      j = (j->result_id() == i.first.id) ? j.Erase() : ++j;
+  }
+
+  // Remove declarations of imported variables
```
These might have OpName attached to them.
Add TODO for their removal.

dneto0 commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+  if (res != SPV_SUCCESS) return res;
+
+  // Rematch import variables/functions to export variables/functions
+  linkedModule->ForEachInst([&linkingsToDo](Instruction* insn) {
```
Suggest you use ReplaceAllUsesWith  ?  But that assumes you have a DefUseManager.  You could create one right now.

dneto0 commented:
Seems to me import/export of functions is not handled. Need a TODO in your list at top, I think

dneto0 commented:
@pierremoreau The build fails on Windows, e.g. https://ci.appveyor.com/project/Khronoswebmaster/spirv-tools/build/1476/job/8crurujh83ltgj1l#L767 
Ambiguous overload

pierremoreau commented on source/link/linker.cpp:
```diff
@@ -0,0 +1,755 @@
+        else if (type == SpvLinkageTypeExport)
+          modulesExports[moduleIndex].emplace(data.name, data);
+      }
+      // TODO(pierremoreau): Can an SpvOpDecorateId be used for
```
Ah, true, I hadn’t properly read the spec about `OpDecorateId`.

pierremoreau commented:
> Seems to me import/export of functions is not handled. Need a TODO in your list at top, I think

It is handled, I even have tests to check that I keep the FunctionCtrl and the FuncParamAttr from the definition, and not from the declaration.

> I've reviewed almost everything again in the main code, and about half the tests.

Thanks a lot for reviewing! That DefUseManager definitely looks interesting,  I’ll look into that.

>        specifically a test for similarity test for OpTypeImage NumInOperands() == 8u case, which I think is a bug.

OpTypeImage has an optional operand, which if set, brings its number of input operands to 8.

>        don't dedup Struct Array, Pointer, RuntimeArray if decorations differ.

I need to change that.

>       fail non-zero schema

What is the instruction schema about? I haven’t found any information on it. However, I should probably fail if any of the modules does not have it equal to 0, to be safe.

>       conflicting generator word: Coders choice about what to do here... :-)

As I use 0 for the generator ID, there are no conflicts in linking modules generated by different tools. There is also a TODO on whether a generator ID should be reserved for the linker or not.

pierremoreau commented on source/opt/basic_block.h:
```diff
@@ -51,6 +57,7 @@ class BasicBlock {
 
   // The label starting this basic block.
   Instruction* GetLabelInst() { return label_.get(); }
+  Instruction GetLabelInst() const { return *label_.get(); }
```
I am unsure why I did not go with returning a constant reference instead of a copy, but I changed it to a constant reference now.

pierremoreau commented on source/opt/instruction.h:
```diff
@@ -58,6 +58,10 @@ struct Operand {
   spv_operand_type_t type;      // Type of this logical operand.
   std::vector<uint32_t> words;  // Binary segments of this logical operand.
 
+  friend bool operator!=(const Operand& o1, const Operand& o2) {
```
I think that is the most common thing to do; I’ll change the code.

pierremoreau commented on source/opt/instruction.h:
```diff
@@ -246,6 +254,39 @@ inline void Instruction::ForEachInst(
+    switch (opnd.type) {
+      case SPV_OPERAND_TYPE_RESULT_ID:
+        f(&opnd.words[0]);
+        result_id_ = opnd.words[0];
```
It needs to set `result_id_` and `type_id_`, as otherwise if you later call `Instruction::result_id()`, you will still get the old value, which is probably not what you would expect as you changed it in the call to `Instruction::ForEachId()`.

However, I do agree that the code could be simplified quite a bit, and look like:
```
for (auto operand : operands_) {
  if (spvIsIdType(operand.type)) f(operand.words[0]);
}
if (result_id_ != 0u) result_id_ = GetSingleWordInOperand(0);
if (type_id_ != 0u) type_id_ = GetSingleWordInOperand(1);
```
And that code should probably check that `f()` did not cause `result_id_` and `type_id_` to equal 0, as that would confuse the `Instruction` class as it relies on them being null to know whether they are set or not.

pierremoreau commented on source/opt/instruction.h:
```diff
@@ -246,6 +254,39 @@ inline void Instruction::ForEachInst(
+}
+
+inline void Instruction::ForEachId(
+    const std::function<void(const uint32_t*)>& f) const {
```
So true! Fixed

pierremoreau commented on source/opt/module.h:
```diff
@@ -66,8 +66,10 @@ class Module {
   inline void AddExecutionMode(std::unique_ptr<Instruction> e);
-  // Appends a debug instruction (excluding OpLine & OpNoLine) to this module.
-  inline void AddDebugInst(std::unique_ptr<Instruction> d);
+  // Appends a debug 1 instruction (excluding OpLine & OpNoLine) to this module.
```
Added. Should I add that comment to all function prototypes containing debug1 or debug2, like `debug1_clear()`?

pierremoreau commented:
You were right for the OpTypeImage NumInOperands() == 8u case, it should be == 7.

pierremoreau commented on source/opt/remove_duplicates_pass.cpp:
```diff
@@ -0,0 +1,225 @@
+             a.GetSingleWordInOperand(5u) == b.GetSingleWordInOperand(5u) &&
+             a.GetSingleWordInOperand(6u) == b.GetSingleWordInOperand(6u) &&
+             a.NumOperands() == b.NumOperands() &&
+             (a.NumInOperands() == 8u ||
```
It should absolutely be.

pierremoreau commented on source/opt/remove_duplicates_pass.cpp:
```diff
@@ -0,0 +1,225 @@
+
+static bool areTypesSimilar(
+    const Instruction& a, const Instruction& b,
+    std::unordered_map<SpvId, const Instruction>& typesMap) {
```
I didn’t now about the DefUseManager; I started using it.

pierremoreau commented on source/opt/remove_duplicates_pass.cpp:
```diff
@@ -0,0 +1,225 @@
+  std::unordered_set<uint32_t> capabilities;
+  for (auto i = module->capability_begin(); i != module->capability_end();) {
+    auto res = capabilities.insert(i->GetSingleWordOperand(0u));
+    i = (res.second) ? ++i : i.Erase();
```
Wait, weren’t you the one that suggested I should use it, in one of your previous reviews? :-p I could very well be misremembering though.
I don’t mind at all switching back to a more readable form.

pierremoreau commented on source/opt/remove_duplicates_pass.cpp:
```diff
@@ -0,0 +1,225 @@
+
+Pass::Status RemoveDuplicatesPass::Process(ir::Module* module) {
+  bool modified = false;
+
```
Yes, that makes sense, will do.

dneto0 commented:
Some answers:
- The schema word:  So far only 0 is allowed.  Presumably it allows us to define a new encoding scheme for instructions in the future, without having to change the header.  Defensive tools should error out if the schema word is other than 0, since in that case they don't really know how to decode anything beyond the header.

- For the generator ID:  The XML registry shows that 0 is reserved by Khronos.   (See https://github.com/KhronosGroup/SPIRV-Headers/blob/master/include/spirv/spir-v.xml).  Maybe the best thing is to reserve a generator ID for the linker.   I've posted a PR to dot hat: https://github.com/KhronosGroup/SPIRV-Headers/pull/46

dneto0 commented:
Sorry about forgetting about import/export functions.  I was tired.  :-)

pierremoreau commented:
* No worries about import/export functions. :-)

* Ah, understood for the schema word! I’ll error out if it is different from 0.

* Thanks for the PR for the generator ID. My understanding was that, even if Khronos reserved 0, it is still allowed for other tools to use it, as some kind of default value: from the SPIR-V spec, section 2.3
  > Generator’s  magic number. It is associated with the tool that generated the module.   Its value does not affect any semantics, and is allowed to be 0.

* FYI, I started working on a basic decoration manager, to help with the code. I addressed some of your comments already, and will address the remaining ones once the decoration manager is working.

dneto0 commented:
* About generator word: I  had forgotten about that!  In any case, the MR to register the linker has been merged, so you can start using 17.  See https://github.com/KhronosGroup/SPIRV-Headers/blob/master/include/spirv/spir-v.xml#L69

* For managing decorations, I remembered there's some good ideas / code in the validator for recording decorations.  See the 4 methods starting at  https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/val/validation_state.h#L231 and the code to scan instructions to build it up: https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/validate_instruction.cpp#L311
   * Bonus points for refactoring that a bit into a reusable bit for both the validator and the linker.  (The optimizer will need something similar.)

pierremoreau commented:
* Thanks for the MR for registering the linker! I saw it was merged, and I’ll update the code to use the proper ID.

* Thanks for the pointers. I haven’t done much on the decoration manager yet, so this will come in handy. I’ll try to finish addressing all your comments and get some working version of the decoration manager for review.

    * Bonus points are interesting, but at the current rate, this merge request will take years to be ready. :-D I’ll see what I can do. :-)

dneto0 commented:
@pierremoreau : I understand this is becoming a rather large piece of work(!).    It's fair if you want to pick a reasonable stopping point with some TODOs, provided they're clearly marked as such.  The tool can be "for experts only" with sharp edges as long as it's clear to users.

pierremoreau commented on source/opt/remove_duplicates_pass.cpp:
```diff
@@ -0,0 +1,225 @@
+
+using spvtools::ir::Instruction;
+
+static bool areTypesSimilar(
```
I added a small comment, and changed Similar to the same. If you have any suggestions on how to improve, I am all ears open.

pierremoreau commented on source/opt/remove_duplicates_pass.cpp:
```diff
@@ -0,0 +1,225 @@
+             (a.NumInOperands() == 8u ||
+              a.GetSingleWordInOperand(7u) == b.GetSingleWordInOperand(7u));
+    case SpvOpTypeSampledImage:
+    case SpvOpTypeRuntimeArray:
```
I added the check for decorations now. Same for the ones below.

pierremoreau commented on source/opt/remove_duplicates_pass.cpp:
```diff
@@ -0,0 +1,225 @@
+
+  // Remove duplicate types
+  // TODO(pierremoreau): optimise this part of the pass
+  const auto isType = [](SpvOp op) {
```
That does the trick, thanks!

pierremoreau commented on test/link/binary_version_test.cpp:
```diff
@@ -0,0 +1,63 @@
+  spvtest::Binaries binaries;
+};
+
+TEST_F(BinaryVersion, Default) {
```
It is marked as a TODO in the linker code, that the current strategy might not be the best one.

pierremoreau commented on test/link/linker_test.h:
```diff
@@ -0,0 +1,102 @@
+    return linker_.Link(binaries, linked_binary, options);
+  }
+
+  spv_result_t Disassemble(const spvtest::Binary& binary, std::string& text) {
```
I do ASSERT_EQ at the call site, but you are right, I should be doing that in the function instead.

pierremoreau replied:
It seems that ASSERT_EQ introduce a void return, so I can’t return anything from the Disassemble function if I assert there. I’ll keep as it is then.

pierremoreau commented on test/link/linker_test.h:
```diff
@@ -0,0 +1,102 @@
+    linker_.SetMessageConsumer(consumer);
+  }
+
+  virtual void SetUp() {}
```
If I skip the TearDown method, if test 1 failed and test 2 fails as well, test 2 will return the error messages from test 1 as well.

pierremoreau commented on test/link/linker_test.h:
```diff
@@ -0,0 +1,102 @@
+  virtual void SetUp() {}
+  virtual void TearDown() { error_message_.clear(); }
+
+  spv_result_t Link(const std::vector<std::string>& bodies, spvtest::Binary& linked_binary, spvtools::LinkerOptions options = spvtools::LinkerOptions()) {
```
I am not sure I would agree with the style guide, as it introduces the possibility of getting a nullptr, and having to check for that, but so be it. :-)

pierremoreau commented on tools/link/linker.cpp:
```diff
@@ -0,0 +1,150 @@
+
+Options:
+  -h, --help       Print this help.
+  -o, --outfile    Name of the resulting linked SPIR-V binary.
```
I’ll go with just -o, as all the other tools in SPIRV-Tools it seems.

pierremoreau commented:
I ended up not using the links you gave me for the first draft of the decoration manager. I’ll look into those while addressing the remaining comments, which should be the ones on `source/link/linker.cpp` and adding additional tests.

pierremoreau commented:
The last series of commits should finish addressing all the comments you made about the code. Now, on to add more tests, finish formatting the code and cleaning up the commit history.

dneto0 commented:
Thanks again for this large contribution, and for your patience.

Don't worry about cleaning up the commit history.  The main thing is to rebase to make sure it applies cleanly.  (I always rebase.)  And since bot testing only matters on the last commit, I will squash everything before pushing into master.

From your TODO list at the top:
- You can skip eliminating duplicate decorations or names.  An optimizer cleanup pass can address that later.  It's still valid to have those duplicates.

Is there anything else you think is needed in order to make a combined module valid?


I'm going to take another pass at this today.

dneto0 commented:
Also, you should make the bots pass.  :-)

If you think it's functionally complete, then remove the WIP from the title so I know it would be ready to merge.

pierremoreau commented:
> Also, you should make the bots pass. :-)

The only one failing one, is due to some test timing out because it is taking too long to run in Debug mode on AppVeyor. It is the one checking for what is happening if the amount value is over the limit, so I end up with a really big amount of global values in my test SPIR-V binaries. Besides optimising the code, I am not too sure on how to solve that.

> If you think it's functionally complete, then remove the WIP from the title so I know it would be ready to merge. 

I haven’t added the additional test files you asked for, so I am keeping the WIP flag for the time being, until I add them; I’ll make sure to do that at the very least this weekend.

> Thanks again for this large contribution, and for your patience.

Thank you for your patience as well, as well as all the reviews and pieces of advice!

> And since bot testing only matters on the last commit, I will squash everything before pushing into master.

I guess some tweaks to the `opt::ir::Module` & co. could be kept apart in separate commits, but I don’t mind squashing all the linker related stuff into one giant commit.

> Is there anything else you think is needed in order to make a combined module valid?

I’ll need to check the code again, but I think everything needed to generate a combined valid module should be there.

dneto0 commented on test/link/binary_version_test.cpp:
```diff
@@ -0,0 +1,54 @@
+  };
+  spvtest::Binary linked_binary;
+
+  ASSERT_EQ(SPV_SUCCESS, Link(binaries, &linked_binary));
```
Use EXPECT_EQ here.
Only use ASSERT_EQ when failure of the check implies you'll crash in the rest of the test.
E.g.    if you are testing a pointer result from a function call, then do ASSERT_EQ(ptr != nullptr) followed by EXPECT_EQ(ptr->blah, ...);
The assert protects the later dereference.

pierremoreau replied:
That one should remain ASSERT_EQ imho, as if it fails, `linked_binary` might be empty, causing the later test `linked_binary[1]` to potentially segfault.

dneto0 commented on test/link/entry_points_test.cpp:
```diff
@@ -0,0 +1,66 @@
+)";
+
+  spvtest::Binary linked_binary;
+  ASSERT_EQ(SPV_SUCCESS, AssembleAndLink({body1, body2}, &linked_binary));
```
Cool.
I didn't realize you could have this case.  Things I learn.  :-)

dneto0 commented on test/link/entry_points_test.cpp:
```diff
@@ -0,0 +1,66 @@
+  spvtest::Binary linked_binary;
+  ASSERT_EQ(SPV_ERROR_INTERNAL,
+            AssembleAndLink({body1, body2}, &linked_binary));
+  EXPECT_THAT(GetErrorMessage(),
```
Nice

dneto0 commented on test/link/global_values_amount_test.cpp:
```diff
@@ -0,0 +1,149 @@
+  spvtest::Binary linked_binary;
+
+  ASSERT_EQ(SPV_ERROR_INTERNAL, Link(binaries, &linked_binary));
+  EXPECT_THAT(GetErrorMessage(),
```
Nice nice nice.

dneto0 commented on test/link/entry_points_test.cpp:
```diff
@@ -0,0 +1,66 @@
+
+  spvtest::Binary linked_binary;
+  ASSERT_EQ(SPV_SUCCESS, AssembleAndLink({body1, body2}, &linked_binary));
+  EXPECT_THAT(GetErrorMessage(), std::string());
```
I haven't seen a pure value used as a matcher.
Things I learn!  :-)

dneto0 commented on test/link/ids_limit_test.cpp:
```diff
@@ -0,0 +1,81 @@
+
+  ASSERT_EQ(SPV_ERROR_INVALID_ID, Link(binaries, &linked_binary));
+  EXPECT_THAT(GetErrorMessage(),
+              HasSubstr("The limit of IDs, 4194303, was exceeded: 4194304 is "
```
Nice

dneto0 commented on test/link/matching_imports_to_exports_test.cpp:
```diff
@@ -0,0 +1,326 @@
+)";
+  std::string res_body;
+  SetDisassembleOptions(SPV_BINARY_TO_TEXT_OPTION_NO_HEADER);
+  ASSERT_EQ(SPV_SUCCESS, Disassemble(linked_binary, &res_body))
```
Use EXPECT_EQ here and on the next line.

dneto0 commented on test/link/matching_imports_to_exports_test.cpp:
```diff
@@ -0,0 +1,326 @@
+  ASSERT_EQ(SPV_ERROR_INVALID_BINARY,
+            AssembleAndLink({body1, body2}, &linked_binary));
+  EXPECT_THAT(GetErrorMessage(),
+              HasSubstr("No export linkage was found for \"foo\"."));
```
I'd probably rephrase this error as   

> Unresolved external reference to "foo"

dneto0 commented on test/link/matching_imports_to_exports_test.cpp:
```diff
@@ -0,0 +1,326 @@
+            AssembleAndLink({body1, body2}, &linked_binary))
+      << GetErrorMessage();
+  EXPECT_THAT(GetErrorMessage(),
+              HasSubstr("Type mismatch between imported variable/function %1 "
```
Please add a TODO to include the linkage name.

(Goodness knows the rest of SPIRV-Tools could improve its messages too...)

dneto0 commented on test/link/matching_imports_to_exports_test.cpp:
```diff
@@ -0,0 +1,326 @@
+      << GetErrorMessage();
+  EXPECT_THAT(
+      GetErrorMessage(),
+      HasSubstr("Too many export linkages, 2, were found for \"foo\"."));
```
Nice.

dneto0 commented on test/link/matching_imports_to_exports_test.cpp:
```diff
@@ -0,0 +1,326 @@
+            AssembleAndLink({body1, body2}, &linked_binary))
+      << GetErrorMessage();
+  EXPECT_THAT(GetErrorMessage(),
+              HasSubstr("Type mismatch between imported variable/function %1 "
```
Nice

dneto0 commented on test/link/matching_imports_to_exports_test.cpp:
```diff
@@ -0,0 +1,326 @@
+)";
+
+  spvtest::Binary linked_binary;
+  ASSERT_EQ(SPV_SUCCESS, AssembleAndLink({body1, body2}, &linked_binary))
```
So the link succeeds even though the function param attributes mismatch?

I don't know enough to say if that's right, but I recommend you rename the test to make it clearer, e.g.
FuncParamAttrDifferButStillMatchExportToImport  (?)

dneto0 commented on test/link/binary_version_test.cpp:
```diff
@@ -0,0 +1,54 @@
+  ASSERT_EQ(SPV_SUCCESS, Link(binaries, &linked_binary));
+  EXPECT_THAT(GetErrorMessage(), std::string());
+
+  ASSERT_EQ(0x00000600u, linked_binary[1]);
```
Use EXPECT_EQ here.

dneto0 commented:
Please rebase to avoid conflicts.

dneto0 commented:
Also, if you weren't so close, I'd suggest splitting off the debug1/debug2 changes into its own PR.  But don't do that now!

dneto0 commented:
Our messages crossed paths.

> I guess some tweaks to the opt::ir::Module & co. could be kept apart in separate commits, but I don’t mind squashing all the linker related stuff into one giant commit.

Makes sense.  Yes, that could go into two different commits.  But there's a race condition if you do the squashing.  If you do the squash then theoretically I'd have to review everything again.   (!)

dneto0 commented:
> The only one failing one, is due to some test timing out because it is taking too long to run in Debug mode on AppVeyor. It is the one checking for what is happening if the amount value is over the limit, so I end up with a really big amount of global values in my test SPIR-V binaries. Besides optimising the code, I am not too sure on how to solve that.

I see.  And I see it's a single test that uses the full 15 minutes.  That would be painful even in every day testing. 

I found and commented on a few places where I saw quadratic behaviour.  Maybe it's one of those?  Please try a little bit to fix it.  (That test takes 11s in Release mode on my machine.)  If necessary we could  guard that test with a special #define and set that define in the bots.

dneto0 commented:
I think you can mark the "remove duplicate decorations" as complete.  :-)

dneto0 commented:
By the way, people are going to love this thing.

It unblocks a bunch of other stuff.  For example, glslc from Shaderc will be able to compile and combine multiple shader translation units on the same command line, i.e. work without the -c option.   And in turn that will enable adding shader compilation to CMake, since CMake wants to treat non-C compilers as if they have normal separate-compilation-and-linking support just like C compilers have.

tl;dr  This is awesome.  Keep going!

dneto0 commented:
There are known non-linearities in instruction and operand lookups in the tables. See https://github.com/KhronosGroup/SPIRV-Tools/issues/502  

If that's the key source of the slowdown then don't fix it; leave it to someone else.  Instead we'll turn off the test in the Appveyor debug config as described above.

dneto0 commented:
I've looked at the long test runtime.  Preliminaries:

- on Linux GCC debug the test takes 10s
- on Windows debug it takes a long time.  I lost patience even with EntryPoints UnderLimit which links an *empty* module:
   - 98% of time (in a truncated run) is spent in DefUseManager::AnalyzeInstDef -> DefUseManager::ClearInst -> EraseUseRecordsOfOperandIds.   

The DefUseManager uses std::list to store uses.  We knew that could get bad, and now it seems to have bitten us.   Also, I'm told that Windows debug iterators all track each other and so you can get pathologically bad performance.

Sooooo:   You can modify the test code so that it's disabled on Windows debug.  Detect the case as having both SPIRV_WINDOWS and _DEBUG defined.
Also add a TODO(dneto) to fix the perf problem.

AWoloszyn commented:
We can optionally turn off tracked iterators in appveyor and keep the test turned on. This may be a nicer option, since it will still notify us of breakages, however we won't get notified of issues with tracked iterators.

It would be a matter of setting _ITERATOR_DEBUG_LEVEL to 0 as a preprocessor define to our compile.
 
I believe setting the environment variables 
```
CXX_FLAGS=-D_ITERATOR_DEBUG_LEVEL=0
C_FLAGS=-D_ITERATOR_DEBUG_LEVEL=0
```
should have the desired effect.

antiagainst commented:
Hmm, actually I submitted https://github.com/KhronosGroup/SPIRV-Tools/pull/764 previously to support `SPIRV_TOOLS_EXTRA_DEFINITIONS`, which is used by DXC to add `_ITERATOR_DEBUG_LEVEL`.

antiagainst commented:
No objections to add `-D_ITERATOR_DEBUG_LEVEL=0` by default.

dneto0 commented:
Summary for @pierremoreau : For this MR you can disable the long-running tests for Debug build on Windows.  We have to tackle scalability problems elsewhere.

pierremoreau commented on README.md:
```diff
@@ -293,6 +311,17 @@ Use option `-h` to print help.
+The linker links multiple SPIR-V binary modules together, resulting in a single
+binary module as output.
+
+This is a work in progress, with the linker not allowing mathematical flags
```
Yes, that is clearer.

pierremoreau commented on include/spirv-tools/linker.hpp:
```diff
@@ -0,0 +1,95 @@
+ private:
+  struct Impl;  // Opaque struct for holding the data fields used by this class.
+  std::unique_ptr<Impl> impl_;  // Unique pointer to implementation data.
+};
```
As you prefer, but it’s true that two free functions would work fine as well.

pierremoreau commented on source/opt/decoration_manager.cpp:
```diff
@@ -0,0 +1,240 @@
+        for (uint32_t i = 1u; i < inst->NumInOperands(); ++i) {
+          if (inst->GetSingleWordInOperand(i) == inst->result_id()) {
+            inst->RemoveInOperand(i);
+            break;
```
That is true… I guess it is no invalid behaviour. Cleaning up decorations could be an additional optimisation pass: removing IDs referenced multiple times in OpGroupDecorate, “ungroup” decorations applied to a single ID through a group.

pierremoreau commented on source/opt/decoration_manager.cpp:
```diff
@@ -0,0 +1,240 @@
+    } else {
+      const auto ids_iter = id_to_decoration_insts_.find(id_to_process);
+      if (ids_iter == id_to_decoration_insts_.end())
+        return std::vector<ir::Instruction*>();
```
I tried re-reading my code to figure out why there was this return (I was thinking we would hit it if there was an error), and found the code super confusing, so I re-wrote it and added some comments.

pierremoreau commented on source/opt/decoration_manager.cpp:
```diff
@@ -0,0 +1,240 @@
+
+std::vector<const ir::Instruction*> DecorationManager::GetDecorationsFor(
+    uint32_t id, bool include_linkage) const {
+  std::vector<const ir::Instruction*> decorations;
```
I put the code in an internal templated and non-const private function, which is called by the two methods, as I couldn’t cast the returned vector to the proper type.

dneto0 commented on source/opt/decoration_manager.cpp:
```diff
@@ -0,0 +1,234 @@
+
+// TODO(pierremoreau): The code will return true for { deco1, deco1 }, { deco1,
+//                     deco2 } when it should return false.
+bool DecorationManager::HaveTheSameDecorations(uint32_t id1,
```
For myself: TODO: This is LeftDecorationsContainedInRightDecorations

dneto0 commented:
Rebased, squashed, patched a bit, and pushed into master as 86627f7b3f7

Great work!

pierremoreau commented:
Thank you a lot for all the reviews and the patience. Sorry for not being very communicative lately, but I’m hoping to contribute again to address your remaining comments, as well as adding missing features and adding more tests.

pierremoreau commented on source/opt/pass_manager.h:
```diff
@@ -63,6 +63,8 @@ class PassManager {
   // corresponding Status::Success if processing is succesful to indicate
   // whether changes are made to the module.
+  //
+  // After running all the passes, they are removed from the list.
```
If you have some situation like:
1. Run some passes;
2. Do something on the code;
3. Run some more passes.
and 2. has to happen after 1. With the previous behaviour, you would have to create a second “PassManager” object as calling “run” on the first one again will run all the passes you already run on top of the new ones.
